examples/kotlin/src/main/kotlin/com/xebia/functional/xef/auto/SynopsisReview.kt

```diff
@@ -0,0 +1,51 @@
+package com.xebia.functional.xef.auto
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Play(val title: String, val era: String)
+
+@Serializable
+data class Synopsis(val summary: String)
+
+@Serializable
+data class Review(val review: String)
+
+@Serializable
+data class Score(val score: Double)
+
+suspend fun main() {
+    ai {
+        val playScore = Play("The power of Zuluastral", "Modern Era")
+            .chain<Play, Synopsis>(this) { play -> synopsisTemplate(play) }
+            .chain<Synopsis, Review>(this) { synopsis -> reviewTemplate(synopsis) }
+            .chain<Review, Score>(this) { review -> scoreTemplate(review) }
```

Comment:
I understand your point and it has been an ongoing discussion on how much prompt "templating/chaining" adds to the basic functionality.

Reply:
There are several issue to consider here from an OSS / Library maintenance point of view.

1. Duplicated APIs, cause confusion for users to use 1 way over the other.
2. More code results in higher maintenance
3. With the current work we're doing in _core_ this brings the question on how we can make this Java friendly.

Reply:
We took a similar decision some time ago, removing chains in favor of being more explicit (or just use function composition!)

Chaining makes more sense in a LangChain-style API, in which you put the results of AI calls under keys in the context. Since all elements of the chain share the context, this is the main way to share data between agents. However, we want a more explicit style, hence the much less powerful design of contexts.

Reply:
I agree with the solution discussed in this thread. Although the chain approach had also influenced me, I comprehend and acknowledge the perspectives shared by Alex and Simon. My apologies, @necosta, for the additional effort needed in this matter. 🙏